Login

gavrilenco opened the pull request.

yanAnciferov committed:
add authorization module on server; add post method /login for authorize; add post method getUserData for get informaton about authorized user

yanAnciferov committed:
add login page, login form, login reducer with validation and action for get token

yanAnciferov committed:
add error handle(if incorrect login or password); redirect on success login; redirect if user is auth and trying to go /registration or /login; logout;

yanAnciferov committed:
add main menu to front; load data about authorize user from server; add route to my page;

yanAnciferov committed:
ON BACKEND: create folder for middleware; create error dispatcher; create middleware for chek db connection; create simple error handler; build link to avatars on select user; refactoring; ON FRONTEND: change destructurization; after logout go to login; add user page; refactoring; dispatch action from lifecycle at profile; fix bugs; go to login page if token was destroyed; show not found component if not found on routing;

gavrilenco commented on back/myapp/constants/common.js:
```diff
@@ -18,6 +15,12 @@ const MAX = "2017-12-31"
 
+const PATH_TO_USER_DATA = "./public/usersData/";
+const SERVER_DOMAINL = "http://localhost:3001";
+const PATH_TO_USER_DATA_C = "/usersData/";
```
PATH_TO_USER_DATA_C what is C?

gavrilenco commented on back/myapp/constants/common.js:
```diff
@@ -18,6 +15,12 @@ const MAX = "2017-12-31"
 const FEMALE = "female"
 
+const PATH_TO_USER_DATA = "./public/usersData/";
+const SERVER_DOMAINL = "http://localhost:3001";
```
SERVER_DOMAINL what is L?

gavrilenco commented on back/myapp/constants/common.js:
```diff
@@ -18,6 +15,12 @@ const MAX = "2017-12-31"
+const SERVER_DOMAINL = "http://localhost:3001";
+const PATH_TO_USER_DATA_C = "/usersData/";
+const FULL_PATH_TO_USER_DATA = SERVER_DOMAINL + PATH_TO_USER_DATA_C;
+const PATH_TO_DEFAULT_AVATAR = SERVER_DOMAINL + PATH_TO_USER_DATA_C + "default_avatar.png"
```
why no to use FULL_PATH_TO_USER_DATA here?

gavrilenco commented on back/myapp/constants/errors.js:
```diff
@@ -17,7 +17,11 @@ const RECT_FIELD_ERROR = "rect fields value must be from 0 to 1"
 const REGISTRATION_USER_MODEL_INVALIDE = " registration user model invalidate"
-
+const UNAUTHORIZED = "unauthorized"
+const INCORECT_DATA_FOR_LOGIN = "incorect data for login"
```
incorect -> incorrect

gavrilenco commented on back/myapp/routes/index.js:
```diff
@@ -22,26 +30,15 @@ var storage = multer.diskStorage({
+router.post(API_METHODS_PATHS.LOGIN, [checkDbConnection, login])
+        .use(simpleErrorHandler);
+        
+router.post(API_METHODS_PATHS.GET_AUTHORIZE_USER_DATA, verifyToken, [checkDbConnection, getAuthUserData])
```
why POST for getting user data?

gavrilenco commented on back/myapp/scripts/account/account.js:
```diff
@@ -0,0 +1,30 @@
+
+
+function getAuthUserData(req, res, next){
+    jwt.verify(req.token, loginConfig.secretKey, (err, authData) => {
```
why we use tokenVerify in other places but here direct call?

gavrilenco commented on back/myapp/scripts/account/login.js:
```diff
@@ -0,0 +1,33 @@
+function login (req,res,next) {
+    var { email, password } = req.body
+
+    if(typeof email == "undefined")
```
if (email) will check email != null && email != undefined

gavrilenco commented on back/myapp/scripts/account/login.js:
```diff
@@ -0,0 +1,33 @@
+
+    User.findOne({email}, (err, user) => {
+      if(user == null){
+        dispatchError(res,next,LOGIN.INCORECT_EMAIL_FOR_LOGIN,403);
```
if (user == null){
dispatch
return
}

if (user.checkPassword)

gavrilenco commented on back/myapp/scripts/errorHandlers/registration.js:
```diff
@@ -0,0 +1,10 @@
+
+function registrationError(err,req,res,next){
+    clearRegistrationRequest(req);
+    res.send(res.message);
```
you have the same logic in simpleErrorHandler, what is for then?

gavrilenco commented on back/myapp/scripts/midllewares/checkDbConnection.js:
```diff
@@ -0,0 +1,14 @@
+    if(mongoose.connection.readyState == 0){
+        res.statusCode = 500;
+        res.message = USER_ERRORS.DB_NOT_CONNECTED;
+        next(res.message)
```
dispatchError?

gavrilenco commented on back/myapp/scripts/midllewares/token.js:
```diff
@@ -0,0 +1,14 @@
+
+function verifyToken(req, res, next){
+    const bearerHeader = req.headers['authorization'];
+    if(typeof bearerHeader !== 'undefined'){
```
if (bearerHeader)

gavrilenco commented on back/myapp/scripts/users.js:
```diff
@@ -0,0 +1,28 @@
+
+
+function getUserData(req, res, next){
+    jwt.verify(req.token, loginConfig.secretKey, (err, authData) => {
```
why not tokenVerify?

gavrilenco commented on back/myapp/scripts/utils.js:
```diff
@@ -28,6 +28,23 @@ function deleteFiles(files){
+function getPathToImages({minAvatar, avatar, _id}){
+  var pathToImage = `${paths.FULL_PATH_TO_USER_DATA}${_id.toString(16)}/` 
+  return {
+    minAvatar: (minAvatar !== null && minAvatar !== undefined) ? pathToImage + minAvatar : paths.PATH_TO_DEFAULT_AVATAR,
```
minAvatar ? code : code

gavrilenco commented on back/myapp/scripts/utils.js:
```diff
@@ -28,6 +28,23 @@ function deleteFiles(files){
+  }
+}
+
+function updateImagePaths(user){
```
not sure this is the best name for this function

gavrilenco commented on back/myapp/scripts/account/login.js:
```diff
@@ -0,0 +1,33 @@
+      if(user == null){
+        dispatchError(res,next,LOGIN.INCORECT_EMAIL_FOR_LOGIN,403);
+      } else if(user.checkPassword(password)) {
+        jwt.sign({user}, loginConfig.secretKey,{ expiresIn: "30 days" }, (err, token) => {
```
FYI: better to have refresh token

gavrilenco commented on back/myapp/scripts/users.js:
```diff
@@ -0,0 +1,28 @@
+      }
+      else {
+        var { id } = req.body;
+        User.findById(id, "firstname surname email avatar minAvatar birthdate parrentname", (err, result) => {
```
the same request with the same field list I've seen before, probably you should keep the list of fields somewhere

yanAnciferov replied:
i think that in future this requests will have different field list

gavrilenco commented on back/myapp/scripts/users.js:
```diff
@@ -0,0 +1,28 @@
+function getUserData(req, res, next){
+    jwt.verify(req.token, loginConfig.secretKey, (err, authData) => {
+      if(err) {
+          dispatchError(res,next,LOGIN.UNAUTHORIZED,403);
```
if (err){
dispatch
return
}

gavrilenco commented on front/social/src/App.js:
```diff
@@ -1,26 +1,33 @@
-import { Route } from "react-router-dom"
+import { Route, Redirect } from "react-router-dom"
 
 import Welcome from "./components/pages/helloWorld.js"
```
Welcome still used on this page?

gavrilenco commented on front/social/src/actions/Account.js:
```diff
@@ -64,4 +66,72 @@ export const registration = () => (dispatch, getState) => {
+
+    dispatch({
+      type: ACTION_FOR_APP.SHOW_LOADING_WINDOW,
+      payload: "Идет авторизация"
```
should be in constants

gavrilenco commented on front/social/src/actions/Account.js:
```diff
@@ -64,4 +66,72 @@ export const registration = () => (dispatch, getState) => {
+
+   if(isAuthorize == false)
+    return;
+    console.log("Получение данных авторизованного пользователя")
```
console.logs should be removed

gavrilenco commented on front/social/src/App.js:
```diff
@@ -1,26 +1,33 @@
     var { errorWindow} = this.props.catcher;
     var { onCloseErrorWindow } = this.props;
+
+    const menu = () => { return (authorizedUser !== null) ? <MainMenu/> : "" };
```
is this a HOC withAuthorize? this logic should be moved to MainMenu component

gavrilenco commented on front/social/src/components/common/header.js:
```diff
@@ -38,16 +43,49 @@ class Header extends React.Component {
-                </Link>
-                <Link to="/registration">
+              {
+                (()=>{
```
looks like it's a separate component (() => {})()

gavrilenco commented on front/social/src/components/common/header.js:
```diff
@@ -38,16 +43,49 @@ class Header extends React.Component {
+                          <CircleAvatar imageUrl={authorizedUser.minAvatar}/>
+                        </div>
+                        <Menu
+                          style={{marginTop:"2.8em"}}
```
inline styles?

gavrilenco commented on front/social/src/components/common/mainMenu.js:
```diff
@@ -0,0 +1,57 @@
+    }
+  
+    render() {
+      const { isAuthorize, authorizedUser } = this.props.app;
```
where is isAuthorize used?

gavrilenco commented on front/social/src/components/pages/profile/avatar.js:
```diff
@@ -0,0 +1,20 @@
+
+    render() {
+      var { user } = this.props;   
+      var forRender = (user !== null) ? <img src={user.minAvatar} /> : <div className="avatar-emitter" />
```
user ? code : code

gavrilenco commented on front/social/src/components/pages/profile/mainInfo.js:
```diff
@@ -0,0 +1,76 @@
+  
+    render() {
+        var { user } = this.props;   
+        var forRender = (user !== null) ? (<MainInfoWithUser user={user}/>) : ( <MainInfoEmitter/>)
```
var why not let or const? is this es6?

gavrilenco commented on front/social/src/reducers/login.js:
```diff
@@ -0,0 +1,98 @@
+        if(data == errors.INCORECT_PASSWORD_FOR_LOGIN)
+            return {
+                ...state,
+                validateState: {
```
looks like we need separate reducer for loginValidate

gavrilenco commented on front/social/src/reducers/catcher.js:
```diff
@@ -10,6 +10,27 @@ const initialState = {
+
+
+function getTechnicalState(state){
+    return {
```
copy paste

gavrilenco commented on front/social/src/components/pages/profile/profile.js:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    update(id){
+      var {props} = this;
```
destructurization?

gavrilenco commented on front/social/src/components/pages/profile/mainInfo.js:
```diff
@@ -0,0 +1,76 @@
+    render() {
+        return (
+            <Paper className="main-info-wrapper">
+                sdfsdf
```
?

gavrilenco commented on front/social/src/components/pages/profile/mainInfo.js:
```diff
@@ -0,0 +1,76 @@
+   render() {
+        var { user } = this.props;
+        var items = [
+            { key: PROFILE_CONTENT.BIRTHDATE, value: new Date(user.birthdate).toLocaleDateString("ru-RU", PROFILE_CONSTS.DATE_FORMAT)}
```
desctructurization?

gavrilenco commented on front/social/src/components/common/mainMenu.js:
```diff
@@ -0,0 +1,57 @@
+                    </MenuItem>
+                </Link>
+                <MenuItem >
+                    <ListItemText primary="Друзья" />
```
constants? how will you implement i18n?

yanAnciferov committed:
global refactoring

yanAnciferov committed:
fixed bugs; auto login after registration; changed backend middleware for ligin and registration; display user sex in profile

yanAnciferov committed:
fixed bugs; refactor form for set user data; start work on edit pdage;

yanAnciferov committed:
add ability edit main user info; start work on update avatar

yanAnciferov committed:
Task: My Account \n\n Backend:\n\n-Fixed bug when editinig email;\n-delete old image files on editing avatars;\n-change virifyToken function;\n-added more functions for removing files; \n\nFrontend\n\n-create AvatarPicker;\n-create SetUserData;\n-refactoring registration pages and registration reduser\n-create wrapper for form fields;\n-create actions for edit user data;\n -create action for update avatar;

yanAnciferov committed:
Task: My Account

Frontend:
-refactored actions/Account.js and actions/Users.js; moved responce and error handlers to other file;
-refactored axios authorization token; created function for update auth header in axios;
-refactored components; moved literals and constants to other files
-removed  unused code

Backend:
-refactored ApiUrl in routers
-db connection check was moved to the global router
